File: C++/programas/Listas-Colas dinamicas/Listas-Colas dinamicas/Cola_Nombres.cpp
#include "Cola_Nombres.h" //Cabecera de clase Cola_Nombres.
#include <queue> //Cabecera para trabajar con colas din?micas Queue.
#include <iostream> //Cabecera para entrada y salida de datos.

//Procedimiento que despliega un men? de opciones para manipular la cola recibida como par?metro.
void Cola_Nombres::Menu(std::queue<std::string> Nombres) {
	int Op; //Variable que recibe la opci?n escogida por el usuario.

	system("cls"); //Limpia la terminal.
	std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    //Despliega el men? de opciones.
	std::cout << "-Cola de nombres-\n\n";
	printf("Nombres registrados: %d\n", Nombres.size());
	std::cout << "\n-Men?-\n1) Insertar nombre.\n2) Eliminar nombre.\n3) Vaciar cola.\n4) Buscar nombre.\n5) Recorrer cola.\n6) Salir.\n";
	//Ingreso de valor a variable.
    std::cout << "Opci?n: "; std::cin >> Op;

    if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
    {
        std::cout << "Error de formato. ";
        system("pause"); //Pausa el flujo del programa.
        std::cin.clear(); //Limpia el buffer de cin.
        //Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        Menu(Nombres); //Llamada recursiva a Menu.
    }
    else
    {
        //Selectiva switch para saltar al m?todo escogido por el usuario.
        switch (Op) {
        case 1:
            AgregarNombre(Nombres);
            break;
        case 2:
            EliminarNombre(Nombres);
            break;
        case 3:
            VaciarCola(Nombres);
            break;
        case 4:
            BuscarNombre(Nombres);
            break;
        case 5:
            RecorrerNombres(Nombres);
            break;
        case 6:
            exit(-1); //Salida del programa
            break;
        default:
            std::cout << "Selecciona una opci?n del men? mostrado.";
            system("pause"); //Pausa el flujo del programa.
            Menu(Nombres); //Llamada recursiva a Menu.
        }
    }
}

//Procedimiento que recorre y despliega los elementos de una cola pasada por par?metro.
void Cola_Nombres::RecorrerNombres(std::queue<std::string> Nombres) {
    int i = 1; //Variable para la iteraci?n de un ciclo.
    std::queue<std::string> temp = Nombres; //Cola temporal para recorrer la cola.

    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    //Si la cola est? vac?a, despliega mensaje.
    if (temp.size() == 0) {
        std::cout << "Cola vac?a. ";
    } else {
        std::cout << "--Alumnos con sus promedios--\n\n";
        while (!temp.empty()) { //Mientras la cola no est? vac?a...
            std::cout << i << " - " << temp.front() << ".\n"; //Despliega el elemento hasta enfrente de la cola.
            temp.pop(); //Elimina el elemento hasta enfrente de la cola y pasa al siguiente.
            i++; //Incrementa contador.
        }
    }

    system("pause"); //Pausa el flujo del programa.
    Menu(Nombres); //Llamada a Menu.
}

//Procedimiento que busca un nombre en la cola pasada por par?metro.
void Cola_Nombres::BuscarNombre(std::queue<std::string> Nombres) {
    std::string nombre; //Variable string que recibe el nombre a buscar.
    int Op; //Variable que recibe opci?n seleccionada por el usuario.
    bool encontrado = false; //Variable auxiliar para la b?squeda del elemento.
    std::queue<std::string> temp = Nombres; //Cola temporal para recorrer la cola.

    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    //Si la cola est? vac?a, despliega mensaje.
    if (Nombres.size() == 0) {
        std::cout << "Cola vac?a. ";
    }
    else {
        std::cout << "Ingresa el nombre a buscar: "; std::cin >> nombre;
        while (!temp.empty()) { //Mientras la cola no est? vac?a...
            if (temp.front().compare(nombre) == 0) { //Compara el elemento de hasta enfrente de la cola con el nombre a buscar.
                encontrado = true;
            }
            temp.pop(); //Elimina el elemento de hasta enfrente de la cola y pasa al siguiente.
        }

        //Si encontr? o no encontr? el elemento, despliega mensaje.
        if (encontrado)
            std::cout << "\nSe encontr? el nombre. ";
        else
            std::cout << "\nNo se encontr? el nombre. ";


        std::cout << "?Desea buscar otro nombre? Si[1]: "; std::cin >> Op;
        if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
        {
            std::cout << "Error de formato. ";
            system("pause"); //Pausa el flujo del programa.
            std::cin.clear(); //Limpia el buffer de cin.
            //Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            Menu(Nombres); //Llamada a Menu.
        }
        else {
            //Dependiendo de la opci?n escogida por el usuario, se llama a m?todo de la clase.
            if (Op == 1)
                BuscarNombre(Nombres);
            else
                Menu(Nombres);
        }
    }

    system("pause"); //Pausa el flujo del programa.
    Menu(Nombres); //Llamada a Menu.
}

//Procedimiento que elimina todos los elementos de la cola pasada por par?metro.
void Cola_Nombres::VaciarCola(std::queue<std::string> Nombres) {
    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    //Si la cola est? vac?a, despliega mensaje.
    if (Nombres.size() == 0) {
        std::cout << "Cola vac?a. ";
    }
    else {
        //Mientras cola cola no est? vac?a, se elimina el elemento al frente de la cola.
        while (!Nombres.empty()) {
            Nombres.pop();
        }

        std::cout << "Cola vaciada. ";
    }

    system("pause"); //Pausa el flujo del programa.
    Menu(Nombres); //Llamada a Menu.
}

//Procedimiento que elimina el elemento al frente de la cola pasada por par?metro.
void Cola_Nombres::EliminarNombre(std::queue<std::string> Nombres) {
    int Op; //Variable que recibe opci?n seleccionada por el usuario.

    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    //Si la cola est? vac?a, despliega mensaje.
    if (Nombres.size() == 0) {
        std::cout << "Cola vac?a. ";
    }
    else {
        std::cout << "Se va a eliminar el primer nombre de la cola. ";
        system("pause"); //Pausa el flujo del programa.
        Nombres.pop(); //Elimina el elemento hasta enfrente de la cola.

        std::cout << "?Desea buscar otro nombre? Si[1]: "; std::cin >> Op;
        if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
        {
            std::cout << "Error de formato. ";
            system("pause"); //Pausa el flujo del programa.
            std::cin.clear(); //Limpia el buffer de cin.
            //Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            Menu(Nombres); //Llamada a Menu.
        }
        else {
            //Dependiendo de la opci?n escogida por el usuario, se llama a m?todo de la clase.
            if (Op == 1)
                EliminarNombre(Nombres);
            else
                Menu(Nombres);
        }
    }

    system("pause"); //Pausa el flujo del programa.
    Menu(Nombres); //Llamada a Menu.
}

//Procedimiento que agrega elementos a la cola pasada por par?metro.
void Cola_Nombres::AgregarNombre(std::queue<std::string> Nombres) {
    int Op; //Variable que recibe opci?n seleccionada por el usuario.
    std::string nombre; //Variable que recibe el nombre a ingresar a la cola.

    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.
    //Ingreso de valor a variable.
    std::cout << "Ingresa un nombre: "; std::cin >> nombre;
    Nombres.push(nombre); //Agrega elemento a la cola.
    std::cout << "Ingreado. ?Desea seguir agregando alumnos? Si[1]: "; std::cin >> Op;

    if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
    {
        std::cout << "Error de formato. ";
        system("pause"); //Pausa el flujo del programa.
        std::cin.clear(); //Limpia el buffer de cin.
        //Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        Menu(Nombres); //Llamada a Menu.
    }
    else
    {
        //Dependiendo de la opci?n escogida por el usuario, se llama a m?todo de la clase.
        if (Op == 1)
            AgregarNombre(Nombres);
        else
            Menu(Nombres);
    }
}

File: Java/programas/Colas dinámicas/src/Cola_Nombres.java
import java.io.IOException; //Excepción si hay error durante entrada y salida de datos.
import java.util.InputMismatchException; //Excepción si hay un error de formato en tipos de datos.
import java.util.Iterator;
import java.util.Queue; //Clase para trabajar con colas dinámicas.
import java.util.Scanner; //Librería para declara objeto Scanner de entrada y salida.

public class Cola_Nombres{
    public Scanner lectura = new Scanner(System.in); //Declara objeto Scanner.

    //Método que despliega un menú para realizar acciones con la cola pasada como argumento de parámetro.
    public void Menu(Queue<String> Nombres) throws InterruptedException, IOException{
        int Op;

        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        System.out.println("-Cola de nombres-\n");
        //Selectiva try-catch por si se ingresa un tipo de dato incorrecto en la variable.
        try{
            System.out.printf("Nombres registrados: %s\n", Nombres.size());
            System.out.println("\n-Menú-\n1) Insertar nombre.\n2) Eliminar nombre.\n3) Vaciar cola.\n4) Buscar nombre.\n5) Recorrer cola.\n6) Salir.");
            System.out.print("Selecciona una opción: "); Op = lectura.nextInt();
            lectura.nextLine(); //Limpia buffer de entrada de datos.
    
            //Selectiva switch para la opción ingresada en el menú.
            switch(Op){
                case 1:
                    AgregarNombre(Nombres);
                    break;
                case 2:
                    EliminarNombre(Nombres);
                    break;
                case 3:
                    VaciarCola(Nombres);
                    break;
                case 4:
                    BuscarNombre(Nombres);
                    break;
                case 5:
                    RecorrerNombres(Nombres);
                    break;
                case 6:
                    lectura.close(); //Cierra el objeto Scanner.
                    System.exit(0); //Cierra el programa.
                    break;
                default:
                    System.out.println("Selecciona una opción del menú mostrado.");
                    lectura.nextLine(); //Limpia buffer de entrada de datos.
                    Menu(Nombres); //Llamada recursiva a Menu.
            }
        } catch(InputMismatchException e){
            System.out.println("Error de formato.");
            lectura.nextLine(); //Limpia buffer de entrada de datos.
            Menu(Nombres); //Llamada recursiva a Menu.
        }
    }

    //Procedimiento que recorrde y despliega todos los elementos de la cola pasada como argumento de parámetro.
    private void RecorrerNombres(Queue<String> nombres) throws InterruptedException, IOException {
        int i;
        Iterator<String> n = nombres.iterator(); //Declara objeto Iterator tipo string para el método iterator()
                                                 //de la cola pasada como parámetro.
        
        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        //Si la cola está vacía, lo despliega y llama a Menu.
        if (nombres.size() == 0){
            System.out.print("Pila vacia. Presione ENTER para continuar...");
        } else{
            System.out.println("---Alumnos con sus promedios---\n");
            for(i = 0; i < nombres.size(); i++){ //Desplega los elementos de la cola.
                System.out.printf("%s - %s\n", i + 1, n.next());
            }
        }
        System.out.print("\nPresione ENTER para regresar al menú...");
        System.in.read(); //Espera a que se presione ENTER para continuar.
        Menu(nombres); //Llamada a Menu.
    }

    //Procedimiento que busca y despliega un elemento de la cola pasada como argumento de parámetro.
    private void BuscarNombre(Queue<String> nombres) throws InterruptedException, IOException {
        String nombre; //Cadena que recibe el nombre a buscar.

        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        //Si la cola está vacía, lo despliega y llama a Menu.
        if (nombres.size() == 0){
            System.out.print("Pila vacia. Presione ENTER para continuar...");
        } else{
            System.out.print("Ingresa el nombre a buscar: "); nombre = lectura.nextLine(); //Solicita nombre a buscar.
            if (nombres.contains(nombre) == true) //Verifica si la cola contiene el nombre a buscar y lo despliega.
                System.out.println("Se encontró el nombre.");
            else
                System.out.print("No se encontró el alumno");
            System.out.print("\nPresione ENTER para regresar al menú...");
        }
        System.in.read(); //Espera a que se presione ENTER para continuar.
        Menu(nombres); //Llamada a Menu.
    }

    //Procedimiento que vacía la cola pasada como argumento de parámetro.
    private void VaciarCola(Queue<String> nombres) throws InterruptedException, IOException {
        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        //Si la cola está vacía, lo despliega y llama a Menu.
        if (nombres.size() == 0){
            System.out.print("Cola vacia. Presione ENTER para continuar...");
        } else{
            nombres.clear(); //Limpia la cola.
            System.out.print("La cola Nombres fue vaciada. Presiona ENTER para regresar al menú...");
        }
        System.in.read(); //Espera a que se presione ENTER para continuar.
        Menu(nombres); //Llamada a Menu.
    }

    //Procedimiento que busca y elimina un elemento de la cola pasada como argumento de parámetro.
    private void EliminarNombre(Queue<String> nombres) throws InterruptedException, IOException {
        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        //Si la cola está vacía, lo despliega y llama a Menu.
        if (nombres.size() == 0){
            System.out.print("Cola vacia. Presione ENTER para continuar...");
        } else{
            System.out.printf("El nombre %s fue eliminado. Presiona ENTER para regresar al menú...", nombres.poll()); //Elimina el objeto en la cima de la cola.
        }
        System.in.read(); //Espera a que se presione ENTER para continuar.
        Menu(nombres);
    }

    //Procedimiento que agrega un elemento de la cola pasada como argumento de parámetro.
    private void AgregarNombre(Queue<String> nombres) throws InterruptedException, IOException {
        int Op;
        String nombre; //Cadena que recibe el nombre a agregar.

        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        //Selectiva try-catch por si se ingresa un tipo de dato incorrecto en la variable.
        try{
            System.out.print("Ingresa un nombre: "); nombre = lectura.nextLine();
            
            nombres.add(nombre); //Agrega elemento String a la cola.
            System.out.print("\nIngresado. ¿Desea seguir agregando alumnos? Si[1]: "); Op = lectura.nextInt();
            lectura.nextLine(); //Limpia buffer de entrada de datos.
            
            //Si se desea agregar otro elemento, hace una llamada recursiva a AgregarNombre, sino, llama a Menu.
            if (Op == 1)
                AgregarNombre(nombres);
            else
                Menu(nombres);
        } catch(InputMismatchException e){
            System.out.println("Error de formato.");
            lectura.nextLine(); //Limpia buffer de entrada de datos.
            AgregarNombre(nombres); //Llamada recursiva a AgregarNombre.
        }
    }
}